src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,436 @@
+        {
+            double f_chi = universal_kepler(chi, r_0, v_0, alpha, delta_t, mu);
+            double df_chi = d_universal_kepler_d_chi(chi, r_0, v_0, alpha, delta_t, mu);
+            delta = f_chi / df_chi;
+            chi = chi - delta;
+            --steps;
+        } while (std::abs(delta) > 1.0E-10 && steps > 1);
+        return chi;
+    }
+
+    double solve_kepler_elliptic(double meanAnomaly, double eccentricity, int steps)
+    {
+        const double tolerance = 1.0E-10;
+        if (std::abs(eccentricity) < 1.0E-9)
+        {
+            return meanAnomaly;
+        }
+
+        double delta;
+        double E = meanAnomaly;
+
+        // use newton's method
+        // todo: implement something which converges more quickly
+        do
+        {
+            double f_E = E - eccentricity * std::sin(E) - meanAnomaly;
+            double df_E = 1.0 - eccentricity * std::cos(E);
+
+            delta = f_E / df_E;
+            E = E - delta;
+
+            E = wrap_angle(E); // Sometimes this sometimes fails to converge without this line
+            --steps;
+        } while (std::abs(delta) > tolerance && steps > 1);
+        return E;
+    }
+
+    double solve_kepler_hyperbolic(double meanAnomaly, double eccentricity, int steps)
+    {
+        const double tolerance = 1.0E-10;
+
+        double delta;
+        double E = std::log(std::abs(meanAnomaly));
+
+        // use newton's method
+        do
+        {
+            double f_E = eccentricity * std::sinh(E) - E - meanAnomaly;
+            double df_E = eccentricity * std::cosh(E) - 1.0;
+
+            delta = f_E / df_E;
+            E = E - delta;
+            --steps;
+        } while (std::abs(delta) > tolerance && steps > 1);
+        return E;
+    }
+
+
+
+
+    double eccentric_anomaly_to_true_anomaly(const double &ecc, const double &E)
+    {
+        // This formula supposedly has better numerical stability than the one below
+        double beta = ecc / (1 + std::sqrt(1 - ecc * ecc));
+        return E + 2.0 * std::atan2(beta * std::sin(E), 1 - beta * std::cos(E));
+
+        // return 2.0 * std::atan2(std::sqrt(1 + ecc) * std::sin(E / 2.0), std::sqrt(1 - ecc) * std::cos(E / 2.0));
+    }
+
+    double hyperbolic_eccentric_anomaly_to_true_anomaly(const double &ecc, const double &H)
+    {
+        return 2 * atan(sqrt((ecc + 1) / (ecc - 1)) * tanh(H / 2));
+    }
+
+    double true_anomaly_to_eccentric_anomaly(const double &ecc, const double &v)
+    {
+        return 2.0 * std::atan(std::sqrt((1 - ecc) / (1 + ecc)) * std::tan(v / 2.0));
+    }
+
+    double true_anomaly_to_hyperbolic_eccentric_anomaly(const double &ecc, const double &v)
+    {
+        return 2.0 * std::atanh(std::clamp(std::sqrt((ecc - 1) / (ecc + 1)) * std::tan(v / 2.0), -1.0, 1.0));
+    }
+
+    double mean_motion(double mu, double a)
+    {
+        return sqrt(mu / (a * a * a));
+    }
+
+    double get_orbiting_radius(double ecc, double semiMajorAxis, double trueAnomaly)
```

Comment:
it looks like all of the functions near the top of this file are internal only to the cpp file.

if that's the case, they should be put into an anonymous namespace, and marked with `static` to help the linker understand they can't "escape" the translation unit.

Reply:
Should it be an anonymous namespace inside of namespace osp, or outside of namespace osp?

Reply:
It doesn't really matter.

Personally I like to put my anonymous namespaces at global scope so that any types from other namespaces are explicitly written out and there's no ambiguity, but other people do it differently.

I don't think we have any guidelines on which to do, so dealer's choice i guess.